router/transformer/transformer.go

```diff
@@ -322,6 +325,223 @@ func (trans *HandleT) makeHTTPRequest(ctx context.Context, url string, payload [
 	return respData, respCode, nil
 }
 
+func (trans *HandleT) MakeDeliveryRequest(ctx context.Context, structData integrations.PostParametersT) (interface{}, error) {
+	postInfo := structData
+	isRest := postInfo.Type == "REST"
+
+	isMultipart := len(postInfo.Files) > 0
+
+	// going forward we may want to support GraphQL and multipart requests
+	// the files key in the response is specifically to handle the multipart usecase
+	// for type GraphQL may need to support more keys like expected response format etc
+	// in future it's expected that we will build on top of this response type
+	// so, code addition should be done here instead of version bumping of response.
+	if isRest && !isMultipart {
+		requestMethod := postInfo.RequestMethod
+		requestBody := postInfo.Body
+		requestQueryParams := postInfo.QueryParams
+		var bodyFormat string
+		var bodyValue map[string]interface{}
+		for k, v := range requestBody {
+			if len(v.(map[string]interface{})) > 0 {
+				bodyFormat = k
+				bodyValue = v.(map[string]interface{})
+				break
+			}
+		}
+
+		var payload io.Reader
+		// support for JSON and FORM body type
+		if len(bodyValue) > 0 {
+			switch bodyFormat {
+			case "JSON":
+				jsonValue, err := jsonfast.Marshal(bodyValue)
+				if err != nil {
+					// panic(err)
+					return nil, fmt.Errorf("400 Unable to parse bodyValue. Unexpected transformer response: %+v", bodyValue)
+				}
+				payload = strings.NewReader(string(jsonValue))
+			case "JSON_ARRAY":
+				// support for JSON ARRAY
+				jsonListStr, ok := bodyValue["batch"].(string)
+				if !ok {
+					trans.logger.Error("400 Unable to parse json list. Unexpected transformer response")
+					return nil, fmt.Errorf("400 Unable to parse json list. Unexpected transformer response")
+					// 	return &utils.SendPostResponse{
+					// 		StatusCode:   400,
+					// 		ResponseBody: []byte("400 Unable to parse json list. Unexpected transformer response"),
+					// 	}
+				}
+				payload = strings.NewReader(jsonListStr)
+			case "XML":
+				strValue, ok := bodyValue["payload"].(string)
+				if !ok {
+					trans.logger.Error("400 Unable to construct xml payload. Unexpected transformer response")
+					return nil, fmt.Errorf("400 Unable to construct xml payload. Unexpected transformer response")
+					// 	return &utils.SendPostResponse{
+					// 		StatusCode:   400,
+					// 		ResponseBody: []byte("400 Unable to construct xml payload. Unexpected transformer response"),
+					// 	}
+				}
+				payload = strings.NewReader(strValue)
+			case "FORM":
+				formValues := url.Values{}
+				for key, val := range bodyValue {
+					formValues.Set(key, fmt.Sprint(val)) // transformer ensures top level string values, still val.(string) would be restrictive
+				}
+				payload = strings.NewReader(formValues.Encode())
+			default:
+				return nil, fmt.Errorf("bodyFormat: %s is not supported", bodyFormat)
+			}
+		}
```

Comment:
Why are we re-implementing all of this logic? We were supposed to reuse the existing logic/module in the router that is already doing it so that we can compare its output with the transformer.

Reply:
https://github.com/rudderlabs/rudder-server/blob/a37ccd97d6bfd6dcf7369ce6afaa7055c2a870bd/router/network.go#L62

Reply:
The idea is to make the proxy request with the prepared destination request and input the delivery payload without sending it to the actual destination.

Reply:
> Why are we re-implementing all of this logic? We were supposed to reuse the existing logic/module in the router that is already doing it so that we can compare its output with the transformer.

1. The internal logic is same as what was there in original(`network.go`)
2. Since it is tightly bound to `SendPost` method, relocated the method to another file for clarity
3. Basically when `transformerProxy` is _enabled_ it actually sends request to `proxyTest` endpoint only & the request will not be sent to the destination at all
4. The destination request( of `http.Request` type) cannot be marshalled directly to easily send it as part of `proxyTest` request, so we had to send the data in a different data-structure so that marshalling takes place &

Reply:
Let's refactor the original code to accommodate our needs. Having duplicate logic even if it is identical would not be ideal.

How about, rather than activating the proxy path, we can use the same router <-> destination path but instead of sending it to the destination, we'd send it to the proxy test route? We can control it with the same config variable if needed.

I feel this is ideal cause we have both the resources (router input payload and destination request) that we can package and send appropriately to the transformer.

Reply:
> Let's refactor the original code to accommodate our needs. Having duplicate logic even if it is identical would not be ideal.

The thing is, why should the original code be disturbed to accommodate something temporary ?
I don't understand why this is not a clean approach to get what we need. Can you explain your perspective around this ?

Reply:
Though these changes sound temporary now, we can reuse them for integration testing later. Having duplicate code is always messy.
The approach I'm suggesting is along the lines of the transformer changes we made (decouple request payload preparation from the transport logic).